Eliminate worker-to-worker communication.

ryzhyk opened the pull request:
TODOs:

- [ ] Handle channel disconnects in both client and worker code
- [x] Run benchmarks
- [x] Rebase on latest master
- [x] Changelog

Kixiron committed:
Removed barriers from within worker runtimes

Removed the redundant `Barrier` (and associated waiting) from within the
timely worker runtimes, this gives us a ~20% speed increase across all
thread counts and brings the three & four worker runtimes down to be
on-par with the two worker runtime.

Results:
```
citations/1 thread/100000 records
time:   [24.575 s 24.941 s 25.489 s]
change: [-15.598% -13.537% -11.100%] (p = 0.00 < 0.05)

citations/2 threads/100000 records
time:   [21.507 s 22.055 s 22.589 s]
change: [-20.123% -17.682% -15.015%] (p = 0.00 < 0.05)

citations/3 threads/100000 records
time:   [22.640 s 22.995 s 23.350 s]
change: [-20.930% -19.293% -17.529%] (p = 0.00 < 0.05)

citations/4 threads/100000 records
time:   [22.302 s 22.736 s 23.154 s]
change: [-22.869% -20.485% -18.435%] (p = 0.00 < 0.05)
```

Kixiron committed:
Distribute updates across workers, removed atomic timestamp

Changed updates to be evenly distributed across workers, this
speeds up large (multi-worker) workloads a good bit (~15%).
The current timely epoch/timestamp is no longer managed by an atomic
integer, but instead each update batch and flush message is tagged with
the timestamp it belongs to (for updates) and the timestamp to advance
to (for flush messages). All workers now have (near) identical runtimes
now, the only difference being a few drops into branches for the leader
(worker 0) for it to send messages back to the coordinator thread, such
as flush acknowledgements. To allow for workers to shut down, an atomic
bool was added which workers check at the beginning of each iteration
of their runtime loop. Additionally, flush messages are now delivered
every worker thread instead of just the leader, allowing timestamps to
stay synchronized across workers.

Results (compared to 0c054e1):

```
citations/1 thread/100000 records
time:   [23.434 s 23.570 s 23.687 s]
change: [-4.642% -5.496% -7.069%] (p = 0.00 < 0.05)

citations/2 threads/100000 records
time:   [18.210 s 18.699 s 19.186 s]
change: [-15.329% -15.216% -15.064%] (p = 0.00 < 0.05)

citations/3 threads/100000 records
time:   [18.973 s 19.525 s 19.983 s]
change: [-16.196% -15.090% -14.419%] (p = 0.00 < 0.05)

citations/4 threads/100000 records
time:   [23.698 s 24.439 s 25.188 s]
change: [+6.259% +7.490% +8.784%] (p = 0.00 < 0.05)
```

Single threaded performance is largely unchanged while 2 & 3 worker
runs see steady improvement. I expect that the number of threads that
we benefit from increases with the volume of data ingested and that 4
workers was a bit too much cross-thread overhead to make up for the 25k
records each received

Kixiron committed:
Load with Acquire ordering and add an extra call to .stop_all_workers()

Kixiron committed:
Broadcast Stop to all workers

Kixiron committed:
Round-robbin workers and drain sent messages

Kixiron committed:
Accumulate commands

Kixiron committed:
Add minimum batch size

ryzhyk committed:
Eliminate worker-to-worker communication.

We fix the protocol between the client and worker threads to avoid race
conditions and redundant unparks.

Our previous design relied on worker-to-worker communication for
progress management with worker 0 being the only one to communicate
directly with client threads. The new implementation in the recent
series of commits by @Kixiron is moving towards a symmetric architecture
with all workers receiving commands and sending acknowledgements.
However it contains some leftovers of the old design in handling of
Flush, and initial data which get broadcast to everyone, but only worker
0 replies to them. This leads to a bunch of unnecessary code (e.g., the
`running` flag and `unpark_peers()`).  In addition, since the worker
synchronization barrier has been removed, it is possible that worker 0
sends the acknowledgement before other workers have finished processing
updates.

To fix these issues, we switch completely to the new model where workers
communicate only with the client thread in `program/mod.rs`.  The client is
responsible for sending commands, unparking all workers and collecting
results from all of them.

- `unpark` is only ever invoked by the client after sending a message to
  a worker.
- Flush, Stop, Query commands are broadcast to all workers.
- All workers send fluch acknowledgements to the client in response to the
  Flush command as well as after processing initial data.
- All worker-to-worker communication, including channels, unpark calls, and
  shared variables, is eliminated.

In addition, this commit reverts two regressions introduced in this series
of changes: replaces `step()` with `step_or_park()` to prevent busy waiting,
and closes non-input session handles after feeding initial data.

ryzhyk committed:
Bump version to 0.37.0.

Kixiron commented on rust/template/differential_datalog/src/program/mod.rs:
```diff
@@ -2262,20 +2280,36 @@ impl RunningProgram {
         })
     }
+
+    fn waitfor_flush_ack(&self) -> Response<()> {
```
Total bikeshed, but I think I prefer a name like `.await_flush()` or `.wait_for_flush()`

Kixiron commented on rust/template/differential_datalog/src/program/worker.rs:
```diff
@@ -136,204 +94,138 @@ impl<'a> DDlogWorker<'a> {
-                        // The sender must unpark worker 0 after sending to the channel.
-                        self.worker.step_or_park(None);
-                    }
+            // After each command received we step, if there's no timely work to be done
+            // our thread will be parked until it's re-awoken by a command
```
```suggestion
            // After each batch of commands received we step, if there's no timely work to be done
            // our thread will be parked until it's re-awoken by a command
```

ryzhyk commented:
The `stream` test is still timing out in CI, and it turns out I did not actually reproduce it.  Not sure if it's stuck or just slow.  Will continue investigating, but not sure how much progress I will make today.

ryzhyk commented:
> The `stream` test is still timing out in CI, and it turns out I did not actually reproduce it. Not sure if it's stuck or just slow. Will continue investigating, but not sure how much progress I will make today.

Correction: the test completed eventually, but it took nearly 50 minutes, which does not seem right.

ryzhyk commented:
I think I nailed down the performance issue. The reason we were not able to reproduce it is that it only shows up when there are many more workers than CPU cores available to DDlog.  Since this particular CI job runs on a public gitlab runner with a single virtual core, it was hit hard even though it only uses 4 workers.  I was able to reproduce this on my laptop using the `taskset` command on Linux to give a single CPU core to DDlog, causing the test to go 100x slower.

The bottleneck is in the crossbeam channel that uses a [spinlock](https://github.com/crossbeam-rs/crossbeam/blob/b71b94412cb1915a0cb2d4ffe043c67b7d935b8a/crossbeam-channel/src/waker.rs#L180) internally. User-level spinlocks are [fundamentally flawed](https://www.realworldtech.com/forum/?threadid=189711&curpostid=189723) and are virtually never a good idea, exactly because they cause this type of superlinear slowdown.  I am shocked that a serious library like crossbeam uses spinlocks. Specifically, several of the channels we use in DDlog are contended: worker-to-main thread channels, profiling channel, and the update handler channel.

Granted, this is a pathological scenario that issues 100K trivial transactions, so its runtime ends up being dominated by crossbeam messaging. But I still find the 100x slowdown disturbing.

I haven't yet investigated if this is a known problem with crossbeam and whether there exist any workarounds.  Will look into this now.  Other takeaways:

- Having an option to disable profiling altogether (or to enable it on demand) is beginning to sound like a good idea to me.
- It would be nice to try building a concurrent update handler, e.g., using dashmap to avoid relying on a global lock or a dedicated thread.

Kixiron commented:
The spinlock is very unsettling, I would have thought crossbeam would use a more intelligent locking strategy (it'd be interesting to see if swapping the spinlock with something like parking_lot's mutex makes a difference). I definitely agree that being able to disable profiling would be a good thing, and I also have a few ideas for leveraging timely in order to make that more efficient, I'll play with it later. For the handling of updates I totally agree, I think actually leveraging timely would be a really great option as opposed to the current strategy, and if done right we could integrate it with channels and even `UnorderedInput`s to allow for completely concurrent transactions, also something I'll play with later

Kixiron commented:
There's also multiple [exponential spinloops](https://github.com/crossbeam-rs/crossbeam/blob/b71b94412cb1915a0cb2d4ffe043c67b7d935b8a/crossbeam-utils/src/backoff.rs), this could be an incredibly pathological case of each thread backing off for increasingly longer periods

ryzhyk commented:
Ok, my assessment was wrong. Spinlocks are silly and show up a lot in the `perf` profile (and we do need better update handler and profiler architecture).  But after a lot more experimentation, it appears that the real problem in this case is in the timely scheduler and it only shows up when the number of workers is larger than the number of cores and also depends on the number of opens session handles.  Here is what I did:

1. Simplified the DDlog program to a bare minimum input -> map -> output. No joins or any other interesting stuff going on. 
2. Disabled all DDlog profiling and update handler.
3. Changed the `apply_update` logic to send all updates to worker 0.

None of these changes made any difference: reducing the number of cores to 1 (when running with 4 workers) slows things down by a factor of 20.

Finally, I add one more change: close all input sessions on all workers, except worker 0, and now the benchmark runs 5x faster, which is 4x slower than on 4 cores.

Conclusion: There is no problem with this PR and I think it can be merged once we've addressed the TODOs. But I want to bring up the scheduling anomaly with @frankmcsherry.  I will try to create a minimal repro and submit an issue in timely.

ryzhyk committed:
Use `step_or_park` instead of `step_while`.

`step_while` causes major performance degradation when the number of
cores available to DD is smaller than the number of workers:
https://github.com/TimelyDataflow/timely-dataflow/issues/362.

`step_or_park` yields the CPU when the worker has no work to do.

Kixiron commented:
That's really obvious now that I think about it. The `.step_while()` change came before I was more familiar with timely's internals, so I was thinking in a purely "what my program does" fashion and didn't know that timely did it's own unparking under the hood which lead to the thought of trying to avoid parking, since every park requires an unpark (which I assumed could only come from us). A really naive bug, that's my bad

ryzhyk commented:
I wish I knew how to code without bugs. It's great that out tests caught it. Interestingly, the stream test was not even designed for this.

ryzhyk committed:
Update Changelog.

ryzhyk commented:
I am going to close this PR and @Kixiron's original PR and create a fresh PR against the main DDlog repo, so I don't need to figure out merge conflicts.
